Allows an engine start from any previous commit

dnhatn opened the pull request:
This commit allows an engine to open with any previous commit, and
replay translog up to a given sequence number. This is a prepared step
to implement for the primary and replica recovery with rollback.

Relates #10708

dnhatn commented:
The PR may not be ready but I open it to discuss. My concern is how to integrate a new introduced `RecoveryConfig` with the existing `OpenMode`. Having two options for a single thing is not good, and `RecoveryConfig` itself is not a good name. On the other hand, using only `OpenMode` requires a large change which is more likely to be rejected.

bleskes commented:
@dnhatn Thanks for picking this up and bringing it for discussion. I'm not fully online but I would like to discuss your thoughts about this approach. I haven't given it some thought in a while now, but if memory serves it will should be enough to delete commits which violate the max seq# >= global checkpoint rule in the onInit of the deletion policy.

Also - are you done with the feedback on the other PRs? I would hope to bring them to closure before discussing any follow up work.

dnhatn commented:
Thanks @bleskes  for your quick feedback. 

> Also - are you done with the feedback on the other PRs?

I have addressed your comments for https://github.com/elastic/elasticsearch/pull/27268 and https://github.com/elastic/elasticsearch/pull/27367. Would you mind taking another look?

> It will should be enough to delete commits which violate the max seq# >= global checkpoint rule in the onInit of the deletion policy

Yes, for a replica, we should open with a proper commit and delete violated commits. However, for a primary we may want to open an engine with the last commit to minimize the number of replaying operations.

bleskes commented:
> I have addressed your comments for #27268 and #27367. Would you mind taking another look?

Sure. 

> Yes, for a replica, we should open with a proper commit and delete violated commits. However, for a primary we may want to open an engine with the last commit to minimize the number of replaying operations.

I don't think the extra operation replay is a big concern - remember that most the times the last commit is good. I like the simplicity of the just doing this in the deletion policy and not making the engine more complex.

dnhatn commented:
> I like the simplicity of the just doing this in the deletion policy and not making the engine more complex.

I agreed. I am closing this as the approach is not good. Thanks @bleskes.
